workflow/ttlcontroller/ttlcontroller.go

```diff
@@ -186,41 +186,57 @@ func (c *Controller) deleteWorkflow(key string) error {
 
 func (c *Controller) ttlExpired(wf *wfv1.Workflow) bool {
 	// We don't care about the Workflows that are going to be deleted, or the ones that don't need clean up.
-	if wf.DeletionTimestamp != nil || wf.Spec.TTLStrategy == nil || wf.Status.FinishedAt.IsZero() {
+	if wf.DeletionTimestamp != nil || (wf.Spec.TTLStrategy == nil && (wf.Status.StoredWorkflowSpec == nil || wf.Status.StoredWorkflowSpec.TTLStrategy == nil)) || wf.Status.FinishedAt.IsZero() {
 		return false
 	}
 	now := c.clock.Now()
-	if wf.Status.Failed() && wf.Spec.TTLStrategy != nil && wf.Spec.TTLStrategy.SecondsAfterFailure != nil {
-		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*wf.Spec.TTLStrategy.SecondsAfterFailure))
+
+	ttlStrategy := wf.Spec.TTLStrategy
+	if ttlStrategy == nil && wf.Status.StoredWorkflowSpec != nil && wf.Status.StoredWorkflowSpec.TTLStrategy != nil {
+		ttlStrategy = wf.Status.StoredWorkflowSpec.TTLStrategy
+	} else {
+		return false
+	}
+
+	if wf.Status.Failed() && ttlStrategy != nil && ttlStrategy.SecondsAfterFailure != nil {
+		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*ttlStrategy.SecondsAfterFailure))
 		return now.After(expiry)
-	} else if wf.Status.Successful() && wf.Spec.TTLStrategy != nil && wf.Spec.TTLStrategy.SecondsAfterSuccess != nil {
-		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*wf.Spec.TTLStrategy.SecondsAfterSuccess))
+	} else if wf.Status.Successful() && ttlStrategy != nil && ttlStrategy.SecondsAfterSuccess != nil {
+		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*ttlStrategy.SecondsAfterSuccess))
 		return now.After(expiry)
 	} else {
-		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*wf.Spec.TTLStrategy.SecondsAfterCompletion))
+		expiry := wf.Status.FinishedAt.Add(time.Second * time.Duration(*ttlStrategy.SecondsAfterCompletion))
 		return now.After(expiry)
 	}
 }
 
 func timeLeft(wf *wfv1.Workflow, since *time.Time) (*time.Duration, *time.Time) {
-	if wf.DeletionTimestamp != nil || wf.Spec.TTLStrategy == nil || wf.Status.FinishedAt.IsZero() {
+	if wf.DeletionTimestamp != nil || (wf.Spec.TTLStrategy == nil && (wf.Status.StoredWorkflowSpec == nil || wf.Status.StoredWorkflowSpec.TTLStrategy == nil)) || wf.Status.FinishedAt.IsZero() {
 		return nil, nil
 	}
+	ttlStrategy := wf.Spec.TTLStrategy
+
+	if ttlStrategy == nil && wf.Status.StoredWorkflowSpec != nil && wf.Status.StoredWorkflowSpec.TTLStrategy != nil {
+		ttlStrategy = wf.Status.StoredWorkflowSpec.TTLStrategy
```

Comment:
Perhaps extract this to a `getTTLStrategy(wf *wfv1.Workflow)` function?

Reply:
`workflow.Spec.GetTTLStrategy()`

Reply:
I prefer a util function `getTTLStrategy(wf *wfv1.Workflow)` because  it needs to check `status` and `spec` for  getting from `TTLStrategy`